File: clients/typescript/switchkeys/api/request/request.ts
import axios, { AxiosRequestConfig, AxiosResponse } from "axios";
import { SwitchKeysLogger } from "../../utils/logger";
import {
  SwitchKeysConnectionError,
  SwitchKeysRecordNotFoundError,
  SwitchKeysUnauthorizedError,
} from "../../core/exceptions";
import SwitchKeysConfig from "../../utils/config";

/**
 * Enum representing HTTP request methods.
 */
enum SwitchKeysRequestMethod {
  POST = "POST",
  GET = "GET",
  PUT = "PUT",
  DELETE = "DELETE",
}

/**
 * Handles making HTTP requests to the API.
 */
class SwitchKeysRequest {
  private logger: SwitchKeysLogger = new SwitchKeysLogger();
  private config = new SwitchKeysConfig();

  /**
   * Makes an HTTP request to the specified URL using the provided method and data.
   * @param url The URL to make the request to.
   * @param method The HTTP method to use for the request.
   * @param data Optional data to send with the request.
   * @returns The response data if the request is successful, or an error message if it fails.
   */
  async call(url: string, method: SwitchKeysRequestMethod, data?: any) {
    let accessToken;
    const requestConfig: AxiosRequestConfig = {
      url,
      method,
      data,
    };

    if (method !== SwitchKeysRequestMethod.GET) {
      accessToken = this.config.load().accessToken;
      requestConfig.headers = {
        Authorization: `Bearer ${accessToken}`,
      };
    }

    try {
      const response = await axios.request(requestConfig);
      return this.readResponse(response);
    } catch (error: any) {
      return this.readResponse(error.response);
    }
  }

  /**
   * Reads the response from the HTTP request and handles errors.
   * @param response The AxiosResponse object containing the response data.
   * @returns The response data if the request is successful, or an error message if it fails.
   */
  private readResponse(response: AxiosResponse) {
    if (!response) {
      throw new SwitchKeysConnectionError(
        "Server might be down or your connection is a bit slow, Please check your connection/server connection."
      );
    }

    const { status, data } = response;

    if (status >= 400 || data.error) {
      return this.displayError(data.message || data.detail, data.error, status);
    } else {
      return this.wrapResults(data.results);
    }
  }

  /**
   * Displays error messages in the console.
   * @param message The error message to display.
   * @param error Optional additional error details.
   */
  private displayError(message: string, error?: any, status?: number) {
    if (status && status === 401) {
      throw new SwitchKeysUnauthorizedError(message);
    }

    if (status && status === 404) {
      throw new SwitchKeysRecordNotFoundError(message);
    }

    if (error) {
      message += `\nError Details:`;
      for (const key of Object.keys(error)) {
        message += `\n- ${key}: ${error[key]}`;
      }
    }

    this.logger.error(message);
  }

  /**
   * Wraps results.
   * @param results The results to wrapped.
   */
  private wrapResults(results: any) {
    return results;
  }
}

export { SwitchKeysRequest, SwitchKeysRequestMethod };

File: clients/python/switchkeys/api/request/request.py
import json
from typing import Any, Dict
import requests
from switchkeys.api.response.response import SwitchKeysResponse
from enum import Enum


class SwitchKeysRequestMethod(Enum):
    POST = "POST"
    GET = "GET"
    PUT = "PUT"
    DELETE = "DELETE"


class SwitchKeysRequest:
    """
    Make an HTTP request to the specified URL using the provided method.

    Args:
        url (str): The URL to make the request to.
        method (SwitchKeysRequestMethod): The HTTP method to use for the request.
        headers (Dict[str, str]): Additional headers for the request.

    Returns:
        SwitchKeysResponse: An object containing the response from the request.

    Raises:
        ValueError: If an invalid method is provided.
    """

    @staticmethod
    def call(
        url: str,
        method: SwitchKeysRequestMethod,
        data: Dict[str, Any] = {},
        token: str | None = None,
    ) -> SwitchKeysResponse:
        try:
            headers = {}
            if token:
                headers["Authorization"] = f"Bearer {token}"

            if method == SwitchKeysRequestMethod.GET:
                response = requests.get(url)
            elif method == SwitchKeysRequestMethod.POST:
                response = requests.post(url, json=data, headers=headers)
            elif method == SwitchKeysRequestMethod.PUT:
                response = requests.put(url, json=data, headers=headers)
            elif method == SwitchKeysRequestMethod.DELETE:
                response = requests.delete(url, headers=headers)
            else:
                raise ValueError("Invalid method provided.")

            response_content = response.content.decode()

            # Check if response content is not valid JSON.
            if response_content.startswith("<!DOCTYPE html>"):
                print("response_content", response_content)
                return SwitchKeysResponse(
                    status_code=response.status_code,
                    error_message="Response content is not valid json.",
                )

            # Check if response content is empty
            if not response_content.strip():
                return SwitchKeysResponse(
                    status_code=response.status_code,
                    error_message="Empty response content",
                )

            response_content = json.loads(response_content)

            if response.status_code >= 200 and response.status_code < 400:
                return SwitchKeysResponse(
                    status_code=response.status_code,
                    message=response_content.get("message"),
                    data=response_content.get("results"),
                )
            elif response.status_code >= 400:
                if response_content.get("detail") is None:
                    error_message = response_content.get("message")
                else:
                    error_message = response_content.get("detail")

                error = response_content.get("error")

                return SwitchKeysResponse(
                    status_code=response.status_code,
                    error_message=error_message,
                    error=error
                )
            else:
                return SwitchKeysResponse(
                    status_code=response.status_code,
                    error_message=response_content.get("detail"),
                )
        except requests.exceptions.RequestException as e:
            return SwitchKeysResponse(
                status_code=500,
                error_message=str(e),
                data=None,
                message=str(e),
            )
